### PR TITLE
Verify sending Content-Type and custom object body via POST

### LINKED ISSUE #997: ClassCastException when POSTing an object without explicitly setting a Content-Type
### Ktor Version

1.1.3

### Ktor Engine Used(client or server and name)

OkHttp with GsonSerializer

### JVM Version, Operating System and Relevant Context

Android 9.0

### Feedback

When I try to do a POST call with a body object, it gives me a ClassCastException:
```
java.lang.ClassCastException: be.gekrabbel.ktortest.MyCustomObject cannot be cast to io.ktor.client.call.HttpClientCall
        at io.ktor.client.HttpClient.execute(HttpClient.kt:151)
        at io.ktor.client.call.HttpClientCallKt.call(HttpClientCall.kt:80)
        at be.gekrabbel.ktortest.MainActivity.callDoesntWork(MainActivity.kt:63)
        at be.gekrabbel.ktortest.MainActivity$onCreate$1.invokeSuspend(MainActivity.kt:41)
```

It works if the body is a String instead of an object, and it also works after explicitly adding `contentType(ContentType.Application.Json)`.

Full sample app: https://github.com/Arne517/ktor-test/blob/master/app/src/main/java/be/gekrabbel/ktortest/MainActivity.kt

### PR DESCRIPTION
**Pull request by e5l**
Close #997, #1127

### CODE REVIEW

**Comment by e5l** [comment]
Hi @arnestockmans, @Ribesg, @lamba92, @mister11. The issue was fixed some time ago, could you take a look and confirm that all cases were covered?

**Comment by e5l** [comment]
@1hsyn1, could you also take a look?

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/SerializationTest.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+class SerializationTest : ClientLoader() {
+
+    @Test
+    fun testSendCustomObject() = clientTests {
+        config {
+            install(JsonFeature)
+        }
+
+        test { client ->
+            assertFailsWith<ClientRequestException> {
```
Comment (Ololoshechkin) [inline ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/SerializationTest.kt:27]: should the exception code (`response.status`) describe the reason of failure (bad content type in header)? If yes -- maybe add the check for it.

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/SerializationTest.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                client.post {
+                    url("https://google.com")
+                    header(HttpHeaders.ContentType, ContentType.Application.Json)
+                    body = MyCustomObject(message = "Hello World")
+                }
+            }
+        }
+    }
+
```
Comment (Ololoshechkin) [inline ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/SerializationTest.kt:36]: Am i correct that 2/2 tests are checking the bad situation (`assertFailsWith`). Shouldn't we add one more test with no failure?

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/SerializationTest.kt (hunk 1 above)
Comment (e5l) [reply ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/SerializationTest.kt:27]: Nope. The message should be default in status
The custom message is usually used for custom status codes

**Comment by Ololoshechkin** [comment]
Also it is not clear to me how this PR solves #997. I think it is required to add checks for content type that was never specified but I couldn't find a test for the situation where no Content-Type header was specified at all.

File: ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/SerializationTest.kt (hunk 2 above)
Comment (e5l) [reply ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/SerializationTest.kt:36]: Will add. Thanks

**Comment by e5l** [comment]
That's true. I'll check that case too.

**Comment by e5l** [comment]
Hey, I fixed the exception message for #997(the behavior is expected, but the message looked confusing) and realized that there are lots of general serialization tests. Could you take a look?

**Pull request merged**